### ISSUE TITLE
`vectorfield_expr` only works with LabelledArrays

### ISSUE THREAD

**Issue by mehalter** [AlgebraicJulia/AlgebraicPetri.jl#137]
We are working on removing LabelledArrays as a package dependency and all of our internal tooling should work with any sort of indexable type for labelled petri nets. While making a PR to remove this dependency from AlgebraicPetri I noticed that the new `vectorfield_expr` outputs a function that strictly relies on LabelledArrays

```julia
using AlgebraicPetri

# Define some random model

sir_lpetri = LabelledPetriNet([:S, :I, :R], :inf => ((:S, :I), (:I, :I)), :rec => (:I, :R))

# Define some parameters and stuff as dictionaries

initial_concentrations = Dict(:S => 990, :I => 10, :R => 0)
model_rates = Dict(:inf => 0.001, :rec => 0.25)

# This works

du = Dict(:S=>0.0, :I=>0.0, :R=>0.0)
out = vectorfield(sir_lpetri)(du, initial_concentrations, model_rates, 0.01)

# This does not

du_expr = Dict(:S=>0.0, :I=>0.0, :R=>0.0)
out = vectorfield_expr(sir_lpetri)(du, initial_concentrations, model_rates, 0.01)

# vectorfield_expr only works with LVectors from LabelledArrays, it should work with Dictionaries/Named Tuples/etc.

using LabelledArrays

du_expr = LVector(;Dict(:S=>0.0, :I=>0.0, :R=>0.0)...)
out = vectorfield_expr(sir_lpetri)(du, LVector(;initial_concentrations...), LVector(;model_rates...), 0.01)
```

@slwu89 pinging since you did this implementation maybe you have a good idea of how to resolve this. It looks like the function relies on the shorthand of like `du_expr[[:S,:I]]` to index multiple keys at one time which we shouldn't be reliant on.

**Comment by slwu89** [comment]
Thanks for the ping @mehalter, I'll look into this today or tomorrow.

**Comment by slwu89** [comment]
@mehalter does the regular `vectorfield` called with `Dict` input for the first arg really work for you? Here I am getting an error, arising from the line `rates = zeros(eltype(du), nt(pn))` in `vectorfield`, because the return of `eltype` for the dict input is a pair.

```
julia> out = vectorfield(sir_lpetri)(du, initial_concentrations, model_rates, 0.01)
ERROR: MethodError: no method matching zero(::Type{Pair{Symbol, Float64}})

Closest candidates are:
  zero(::Union{Type{P}, P}) where P<:Dates.Period
   @ Dates ~/.julia/juliaup/julia-1.9.0+0.x64.linux.gnu/share/julia/stdlib/v1.9/Dates/src/periods.jl:51
  zero(::ForwardDiff.Partials)
   @ ForwardDiff ~/.julia/packages/ForwardDiff/vXysl/src/partials.jl:39
  zero(::Catlab.ACSetInterface.ACSet)
   @ Catlab ~/.julia/packages/Catlab/T3JuV/src/categorical_algebra/Limits.jl:361
  ...

Stacktrace:
 [1] zeros(#unused#::Type{Pair{Symbol, Float64}}, dims::Tuple{Int64})
   @ Base ./array.jl:585
 [2] zeros(#unused#::Type{Pair{Symbol, Float64}}, dims::Int64)
   @ Base ./array.jl:580
 [3] (::AlgebraicPetri.var"#f#23"{LabelledPetriNet, Matrix{Int64}, TransitionMatrices})(du::Dict{Symbol, Float64}, u::Dict{Symbol, Int64}, p::Dict{Symbol, Float64}, t::Float64)
   @ AlgebraicPetri ~/Desktop/git/AlgebraicPetri.jl/src/AlgebraicPetri.jl:301
 [4] top-level scope
   @ REPL[11]:1

julia> eltype(du)
Pair{Symbol, Float64}
```

**Comment by mehalter** [comment]
Yeah, I removed the `eltype` from the definition in that commit. Make sure you have the latest revision of the branch. But the tests and everything are passing on that branch :D

**Comment by mehalter** [comment]
Here is the latest `vectorfield`

https://github.com/AlgebraicJulia/AlgebraicPetri.jl/blob/d4e30ffae1fabff8733dbb8410e9191d212110fb/src/AlgebraicPetri.jl#L295-L310

**Comment by slwu89** [comment]
Ah I see now, just updated my branch, thanks!

**Comment by mehalter** [comment]
This is resolved with #139, thanks again @slwu89 !
